### PR TITLE
Fix duplicate tags after inserting custom tags in the TagBox (T690840)

### PR DESCRIPTION
**Pull request by KsyushaM**

### CODE REVIEW

File: js/ui/tag_box.js [hunk 1]
```diff
@@ -836,20 +836,27 @@ var TagBox = SelectBox.inherit({
...
-            cache = {};
+        var items = [];
+        var cache = {};
+        var thisValueExpr = this._valueGetterExpr() === "this";
+        var filteredValues = {};
 
-        each(values, function(valueIndex, value) {
-            var item = filteredItems[valueIndex];
+        for(var i = 0; i < filteredItems.length; i++) {
+            filteredValues[thisValueExpr ? JSON.stringify(filteredItems[i]) : this._valueGetter(filteredItems[i])] = filteredItems[i];
```
Comment (DokaRus) [inline js/ui/tag_box.js]: Can you simplify this expression?

File: js/ui/tag_box.js [hunk 2]
```diff
@@ -836,20 +836,27 @@ var TagBox = SelectBox.inherit({
...
+        }
+
+        for(var i = 0; i < values.length; i++) {
+            var value = values[i];
+            var currentItem = filteredValues[thisValueExpr ? JSON.stringify(value) : value];
```
Comment (DokaRus) [inline js/ui/tag_box.js]: Perhaps it would be better to extract key as a separate variable

File: js/ui/tag_box.js [hunk 3]
```diff
@@ -836,20 +836,27 @@ var TagBox = SelectBox.inherit({
     },
 
     _createTagsData: function(values, filteredItems) {
-        var items = [],
-            cache = {};
+        var items = [];
+        var cache = {};
+        var thisValueExpr = this._valueGetterExpr() === "this";
```
Comment (churkin) [inline js/ui/tag_box.js]: `isValueExprSpecified` ?

File: js/ui/tag_box.js [hunk 4]
```diff
@@ -836,20 +836,27 @@ var TagBox = SelectBox.inherit({
...
+        var filteredValues = {};
 
-        each(values, function(valueIndex, value) {
-            var item = filteredItems[valueIndex];
+        for(var i = 0; i < filteredItems.length; i++) {
```
Comment (churkin) [inline js/ui/tag_box.js]: use `filteredItems.forEach`

File: js/ui/tag_box.js [hunk 5]
```diff
@@ -836,20 +836,27 @@ var TagBox = SelectBox.inherit({
...
+                    this._createTagData(items, item, value, i);
                 }).bind(this));
             } else {
-                this._createTagData(items, item, value, valueIndex);
+                this._createTagData(items, currentItem, value, i);
             }
-        }.bind(this));
+        }
 
         return items;
```
Comment (churkin) [inline js/ui/tag_box.js]: Is this `async` function?

File: testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js [hunk 6]
```diff
@@ -3206,6 +3206,23 @@ QUnit.test("adding the custom tag should clear input value (T385448)", function(
     assert.equal($input.val(), "", "the input is empty");
 });
 
+QUnit.test("adding the custom tag shouldn't lead to duplicating of ordinary tags ", function(assert) {
```
Comment (churkin) [inline testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js]: ES6

File: testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js [hunk 7]
```diff
@@ -3206,6 +3206,23 @@ QUnit.test("adding the custom tag should clear input value (T385448)", function(
...
+        $input = $tagBox.find("input");
+
+    keyboardMock($input)
+        .type("custom")
+        .press("enter");
+
+    $($tagBox.find("." + LIST_ITEM_CLASS).first()).trigger("dxclick");
+    var $tags = $tagBox.find(".dx-tag");
+
+    assert.equal($tags.length, 2, "only two tags are added");
```
Comment (churkin) [inline testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js]: `strictEqual`

Comment (churkin) [inline testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js]: Please check tag names also

File: testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js [hunk 8]
```diff
@@ -3206,6 +3206,23 @@ QUnit.test("adding the custom tag should clear input value (T385448)", function(
...
+            acceptCustomValue: true,
+            items: [1, 2, 3]
+        }),
+        $input = $tagBox.find("input");
+
+    keyboardMock($input)
+        .type("custom")
+        .press("enter");
+
+    $($tagBox.find("." + LIST_ITEM_CLASS).first()).trigger("dxclick");
```
Comment (churkin) [inline testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js:3221]: Why we should generate `dxclick` here?

**Review by churkin — CHANGES_REQUESTED** [review]

File: js/ui/tag_box.js (hunk 3 above)
Comment (KsyushaM) [reply js/ui/tag_box.js]: Very likely

File: js/ui/tag_box.js (hunk 5 above)
Comment (KsyushaM) [reply js/ui/tag_box.js]: Yes

File: js/ui/tag_box.js (hunk 4 above)
Comment (KsyushaM) [reply js/ui/tag_box.js]: What's the point?

File: testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js (hunk 8 above)
Comment (KsyushaM) [reply testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js:3221]: We need to pick the ordinary tag from the list like [here](https://isc.devexpress.com/Thread/WorkplaceDetails/T690840)

File: js/ui/tag_box.js (hunk 5 above)
Comment (churkin) [reply js/ui/tag_box.js]: You should return a promise then.

File: js/ui/tag_box.js (hunk 4 above)
Comment (churkin) [reply js/ui/tag_box.js]: It's more readable and simpler

**Review by churkin — CHANGES_REQUESTED** [review]

File: testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js (hunk 8 above)
Comment (churkin) [reply testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js:3221]: ok

**Review by DokaRus — DISMISSED** [review]

**Review by DokaRus — APPROVED** [review]

**Review by churkin — APPROVED** [review]

**Pull request merged**